src/transformers/integrations.py

```diff
@@ -961,138 +963,110 @@ class NeptuneCallback(TrainerCallback):
 
-    def on_log(self, args, state, control, logs, model=None, **kwargs):
-        if not self._initialized:
-            self.setup(args, state, model)
-        if state.is_world_process_zero:
-            logs = rewrite_logs(logs) # Note that this is implemented in transformers.integrations.
-            logs = self._add_base_namespace(logs)
-            for k, v in logs.items():
-                self._neptune_run[k].log(v, step=state.global_step)
-                
-    def on_train_end(self, args, state, control, **kwargs):
-        if self._initialized and state.is_world_process_zero:
-            # Log checkpoints
-            if self._log_checkpoints == 'last' and self._neptune_run is not None:
-                # This is hacky, and in some cases may not work as expected. 
-                # For example, if some of the checkpoints in this directory are from an old run, 
-                # they may still get uploaded. Also, I assume that the path structure of the saved checkpoints
-                # is always the same.
-                ind = -1
-                for f in os.listdir(args.output_dir):
-                    if f.startswith("checkpoint-"):
-                        ind_new = int( f.split("-")[1] )
-                        if ind_new > ind:
-                            ind = ind_new
-                if ind > -1:
-                    self._upload_checkpoint(args.output_dir, ind)
-                
-    def on_save(self, args, state, control, **kwargs):
-        """
-        Upload the checkpoints when it is saved to the hard drive by the trainer. 
-        `log_checkpoints` must be set to `'same'`.
-        """
-        if self._log_checkpoints == 'same' and self._neptune_run is not None:
-            self._upload_checkpoint(args.output_dir, state.global_step)
-            
     def __del__(self):
-        """
-        Environment:
-            NEPTUNE_STOP_TIMEOUT (`int`, *optional*):
-                Number of seconsds to wait for all Neptune.ai tracking calls to finish, before stopping the tracked
-                run. If not set it will wait for all tracking calls to finish.
-        """
-        try:
-            stop_timeout = os.getenv("NEPTUNE_STOP_TIMEOUT")
-            stop_timeout = int(stop_timeout) if stop_timeout else None
-            self._neptune_run.stop(seconds=stop_timeout)
-        except AttributeError:
-            pass
+        self._stop_run_if_exists()
+
+    def on_train_begin(self, args, state, control, model=None, **kwargs):
+        self._initialize_run()
+
+    @classmethod
+    def get_run(cls, trainer):
+        for callback in trainer.callback_handler.callbacks:
+            if isinstance(callback, cls):
+                return callback.run
+
+        # TODO: NPT-12189 Update the exception
+        raise Exception("Trainer has any NeptuneCallback configured")
+
+    def on_log(self, args, state, control, logs: Dict[str, float] = None, **kwargs):
+        for k, v in logs.items():
```

Comment:
In engineering specs you've checked if run is initialized here by checking `if self._run:`.

I think you should also test if logs is `None` as well, otherwise you'll get exception on `logs.items()`.

Reply:
Good point! Added